client/go/cmd/logfmt/handleline.go

```diff
@@ -5,58 +5,120 @@
 package logfmt
 
 import (
+	"bytes"
+	"encoding/json"
 	"fmt"
 	"strconv"
 	"strings"
 	"time"
 )
 
-// handle a line in "vespa.log" format; do filtering and formatting as specified in opts
+type logFields struct {
+	timestamp string // seconds, optional fractional seconds
+	host      string
+	pid       string // pid, optional tid
+	service   string
+	component string
+	level     string
+	messages  []string
+}
 
-func handleLine(opts *Options, line string) (output string, err error) {
-	fields := strings.SplitN(line, "\t", 7)
-	if len(fields) < 7 {
+// handle a line in "vespa.log" format; do filtering and formatting as specified in opts
+func handleLine(opts *Options, line string) (string, error) {
+	fieldStrings := strings.SplitN(line, "\t", 7)
+	if len(fieldStrings) < 7 {
 		return "", fmt.Errorf("not enough fields: '%s'", line)
 	}
-	timestampfield := fields[0] // seconds, optional fractional seconds
-	hostfield := fields[1]
-	pidfield := fields[2] // pid, optional tid
-	servicefield := fields[3]
-	componentfield := fields[4]
-	levelfield := fields[5]
-	messagefields := fields[6:]
+	fields := logFields{
+		timestamp: fieldStrings[0],
+		host:      fieldStrings[1],
+		pid:       fieldStrings[2],
+		service:   fieldStrings[3],
+		component: fieldStrings[4],
+		level:     fieldStrings[5],
+		messages:  fieldStrings[6:],
+	}
 
-	if !opts.showLevel(levelfield) {
+	if !opts.showLevel(fields.level) {
 		return "", nil
 	}
-	if opts.OnlyHostname != "" && opts.OnlyHostname != hostfield {
+	if opts.OnlyHostname != "" && opts.OnlyHostname != fields.host {
 		return "", nil
 	}
-	if opts.OnlyPid != "" && opts.OnlyPid != pidfield {
+	if opts.OnlyPid != "" && opts.OnlyPid != fields.pid {
 		return "", nil
 	}
-	if opts.OnlyService != "" && opts.OnlyService != servicefield {
+	if opts.OnlyService != "" && opts.OnlyService != fields.service {
 		return "", nil
 	}
-	if opts.OnlyInternal && !isInternal(componentfield) {
+	if opts.OnlyInternal && !isInternal(fields.component) {
 		return "", nil
 	}
-	if opts.ComponentFilter.unmatched(componentfield) {
+	if opts.ComponentFilter.unmatched(fields.component) {
 		return "", nil
 	}
-	if opts.MessageFilter.unmatched(strings.Join(messagefields, "\t")) {
+	if opts.MessageFilter.unmatched(strings.Join(fields.messages, "\t")) {
 		return "", nil
 	}
 
+	switch opts.Format {
+	case FormatRaw:
+		return line + "\n", nil
+	case FormatJSON:
+		return handleLineJson(opts, &fields)
+	case FormatVespa:
+		fallthrough
+	default:
+		return handleLineVespa(opts, &fields)
+	}
+}
+
+func parseTimestamp(timestamp string) (time.Time, error) {
+	secs, err := strconv.ParseFloat(timestamp, 64)
+	if err != nil {
+		return time.Time{}, err
+	}
+	nsecs := int64(secs * 1e9)
+	return time.Unix(0, nsecs), nil
+}
+
+func handleLineJson(_ *Options, fields *logFields) (string, error) {
+	timestamp, err := parseTimestamp(fields.timestamp)
+	if err != nil {
+		return "", err
+	}
+	outputFields := struct {
```

Comment:
I would prefer this struct to be defined (with a name) outside the function, but that's just a minor style thing.